awswrangler/s3/_select.py

```diff
@@ -0,0 +1,216 @@
+    >>> df = wr.s3.select_query(
+    ...     sql='SELECT * FROM s3object[*][*]',
+    ...     path='s3://bucket/key.json.gzip',
+    ...     input_serialization='JSON',
+    ...     input_serialization_params={
+    ...         'Type': 'Document',
+    ...     },
+    ...     compression="gzip",
+    ... )
+
+    Reading an entire CSV object using threads
+
+    >>> import awswrangler as wr
+    >>> df = wr.s3.select_query(
+    ...     sql='SELECT * FROM s3object',
+    ...     path='s3://bucket/key.csv',
+    ...     input_serialization='CSV',
+    ...     input_serialization_params={
+    ...         'FileHeaderInfo': 'Use',
+    ...         'RecordDelimiter': '\r\n'
+    ...     },
+    ...     use_threads=True,
+    ... )
+
+    Reading a single column from Parquet object with pushdown filter
+
+    >>> import awswrangler as wr
+    >>> df = wr.s3.select_query(
+    ...     sql='SELECT s.\"id\" FROM s3object s where s.\"id\" = 1.0',
+    ...     path='s3://bucket/key.snappy.parquet',
+    ...     input_serialization='Parquet',
+    ... )
+    """
+    if path.endswith("/"):
+        raise exceptions.InvalidArgumentValue("<path> argument should be an S3 key, not a prefix.")
+    if input_serialization not in ["CSV", "JSON", "Parquet"]:
+        raise exceptions.InvalidArgumentValue("<input_serialization> argument must be 'CSV', 'JSON' or 'Parquet'")
+    if compression not in [None, "gzip", "bzip2"]:
+        raise exceptions.InvalidCompression(f"Invalid {compression} compression, please use None, 'gzip' or 'bzip2'.")
+    if compression and (input_serialization not in ["CSV", "JSON"]):
+        raise exceptions.InvalidArgumentCombination(
+            "'gzip' or 'bzip2' are only valid for input 'CSV' or 'JSON' objects."
+        )
+    bucket, key = _utils.parse_path(path)
+
+    args: Dict[str, Any] = {
+        "Bucket": bucket,
+        "Key": key,
+        "Expression": sql,
+        "ExpressionType": "SQL",
+        "RequestProgress": {"Enabled": False},
+        "InputSerialization": {
+            input_serialization: input_serialization_params,
+            "CompressionType": compression.upper() if compression else "NONE",
+        },
+        "OutputSerialization": {
+            "JSON": {},
+        },
+    }
+    if s3_additional_kwargs:
+        args.update(s3_additional_kwargs)
+    _logger.debug("args:\n%s", pprint.pformat(args))
+
+    if any(
+        [
+            compression,
+            input_serialization_params.get("AllowQuotedRecordDelimiter"),
+            input_serialization_params.get("Type") == "Document",
+        ]
+    ):  # Scan range is only supported for uncompressed CSV/JSON, CSV (without quoted delimiters)
+        # and JSON objects (in LINES mode only)
+        _logger.debug("Scan ranges are not supported given provided input.")
+        client_s3: boto3.client = _utils.client(service_name="s3", session=boto3_session)
+        return _select_object_content(args=args, client_s3=client_s3)
```

Comment:
If you cannot use scan ranges, then make a direct call to S3 select (entire object size). No threading possible